### PR TITLE
KAFKA-16570 FenceProducers API returns "unexpected error" when succes…

### PR DESCRIPTION
**Pull request by edoardocomar**
…sful

* Client handling of ConcurrentTransactionsException
* Unit test
* Integration test

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/admin/internals/FenceProducersHandler.java [hunk 1]
```diff
@@ -101,7 +101,8 @@ public ApiResult<CoordinatorKey, ProducerIdAndEpoch> handleSingleResponse(
         InitProducerIdResponse response = (InitProducerIdResponse) abstractResponse;
 
         Errors error = Errors.forCode(response.data().errorCode());
-        if (error != Errors.NONE) {
+        // Concurrent Transaction is still a success for fencing old producers
```
Comment (jolshan) [inline clients/src/main/java/org/apache/kafka/clients/admin/internals/FenceProducersHandler.java]: nit: when you say old producers here, you just mean the fenced producer?

File: core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  private var adminClient: Admin = _
+  private var producer: KafkaProducer[Array[Byte], Array[Byte]] = _
+
+  @BeforeEach
+  override def setUp(testInfo: TestInfo): Unit = {
+    super.setUp(testInfo)
+
+    val producerProps = new Properties
+    producerProps.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, txnId)
+    producerProps.put(ProducerConfig.TRANSACTION_TIMEOUT_CONFIG, 200)
```
Comment (jolshan) [inline core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala]: do we want to set this value low? I don't think we want the timeout to potentially interfere with the test.

File: core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  def overridingProps(): Properties = {
+    val props = new Properties()
+    props.put(ServerLogConfigs.AUTO_CREATE_TOPICS_ENABLE_CONFIG, false.toString)
+    // Set a smaller value for the number of partitions for
+    props.put(TransactionLogConfigs.TRANSACTIONS_TOPIC_PARTITIONS_CONFIG, 1.toString)
+    props.put(TransactionLogConfigs.TRANSACTIONS_TOPIC_REPLICATION_FACTOR_CONFIG, 1.toString)
+    props.put(TransactionLogConfigs.TRANSACTIONS_TOPIC_MIN_ISR_CONFIG, 1.toString)
+    props.put(TransactionStateManagerConfigs.TRANSACTIONS_ABORT_TIMED_OUT_TRANSACTION_CLEANUP_INTERVAL_MS_CONFIG, "200")
```
Comment (jolshan) [inline core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala]: ditto to this one.

File: clients/src/main/java/org/apache/kafka/clients/admin/internals/FenceProducersHandler.java (hunk 1 above)
Comment (edoardocomar) [reply clients/src/main/java/org/apache/kafka/clients/admin/internals/FenceProducersHandler.java]: yes

File: core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala (hunk 2 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala]: I borrowed the value from kafka.api.TransactionsTest where there are fencing tests

File: core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala (hunk 3 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala]: I borrowed the value from kafka.api.TransactionsTest where there are fencing tests

**Comment by edoardocomar** [comment]
Hi @jolshan thanks for taking a look. I added a 2nd integration test (fence after commit) and that proved to be flaky when run continuously. 
The conclusion is that the concurrent  transactions error  cannot be swallowed, must be retried. I am updating the PR

**Comment by jolshan** [comment]
Hmmm -- did you push the test that was flaky?

**Comment by edoardocomar** [comment]
no i pushed the more complete integration test together with the new fix for the client handler

File: core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                   "Unexpected ExecutionException cause " + ee.getCause.getCause)
+      case e: Exception =>
+        throw e
+    }
+
+    try {
+      producer.commitTransaction()
+      fail("expected Exception")
+    } catch {
+      case _: ProducerFencedException =>
```
Comment (jolshan) [inline core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala:134]: any reason why we specify these exceptions if they all have the same behavior?

**Comment by jolshan** [comment]
Hmmm -- so the behavior is now to keep retrying until the transaction actually is aborted. Just curious -- I would expect not to need a retry so I'm curious about why the test was failing.

**Comment by edoardocomar** [comment]
with the client handler ignoring the CONCURRENT_TRANSACTIONS error 
it happened often enough that after the admin.fence call, the producer which should have been fenced could carry on.
e.g. if I restore that client behavior I get :

```
  ...
    producer.commitTransaction()

    adminClient.fenceProducers(Collections.singletonList(txnId)).all().get()

    producer.beginTransaction()
    try {
        producer.send(record).get()
        fail("expected ProducerFencedException") // <- test failed here
    } catch {
```

**Comment by edoardocomar** [comment]
Hi @jolshan would you like any further changes ?

**Comment by jolshan** [comment]
>  producer.beginTransaction()
    try {
        producer.send(record).get()
        fail("expected ProducerFencedException") // <- test failed here
     
 Interesting. I thought it would fence immediately but we might see some extra concurrent transaction errors. I'm curious if we could commit such a transaction in this state, but maybe this is good enough.
 
 The only other thing I could think of is will a long blocking call be an issue for users of this api. Curious about @C0urante's opinion since I believe he was the one who implemented this for connect.

**Comment by edoardocomar** [comment]
> I'm curious if we could commit such a transaction in this state, but maybe this is good enough.

@jolshan  

Yes, easy to test.
If I reinstate the client handling of ignoring the concurrent txn error, 
and remove the assertions from the test , then I can run 
```
  def testFenceAfterProducerCommit(quorum: String): Unit = {
    producer.initTransactions()
    producer.beginTransaction()
    producer.send(record).get()
    producer.commitTransaction()

    adminClient.fenceProducers(Collections.singletonList(txnId)).all().get()

    producer.beginTransaction()
    producer.send(record).get()
    producer.commitTransaction()
  }
```
and it 'passes', i.e. completes without exceptions, very often in zk mode and occasionally in kraft.

**Comment by edoardocomar** [comment]
@C0urante any views ? 
@gharris1727 ?
I'd like this bug to be fixed in 3.8.0 and 3.7.1

**Comment by C0urante** [comment]
It could be a problem for Connect if calls to `Admin::fenceProducers` block, but if the blocking takes place in the returned future instead of the method invocation itself, then that should be fine.

The behavior demonstrated in the integration tests looks good to me, but I defer to Justine unless there's a question of how this API is used in Connect.

**Comment by edoardocomar** [comment]
Thanks @C0urante - the block is in waiting for the future completion.
The problem with the unexpected failure with concurrent txn, is that it leaves a Connect Task in failed state requiring an external restart.

**Comment by C0urante** [comment]
Which unexpected failure are you referring to? One that exists today and is addressed by this change?

If a task's producer fails to commit a transaction (or to send a record during that transaction) because `Admin::fenceProducers` was called, this is completely fine. It should not require a manual restart of the task because the only time this should happen is when the task should already be shut down.

If a worker creates an admin client and invokes `Admin::fenceProducers`, then any failures in that invocation will cause tasks to fail, and we should try to make the method as resilient as possible to retriable errors (and, with my limited understanding of this part of the code base, it seems like that should include `CONCURRENT_TRANSACTIONS`).

**Comment by edoardocomar** [comment]
Yes, I meant the currently possible failure in invoking `Admin::fenceProducers` throwing `CONCURRENT_TRANSACTIONS`.
that this PR addresses.

when running eg mirrormaker2 in exactly once mode, sooner or later this error fails a task . 

I believe this PR makes connect in exactly once mode more robust

**Comment by edoardocomar** [comment]
@jolshan ... gentle nudge for approval or objections ... 
3.8.0 deadline is tomorrow June 12th EOD.

**Review by C0urante — APPROVED** [review]
LGTM

If @jolshan has any objections (especially if this would somehow lead to a regression), we can revert and discuss further. But from what I can tell this is a safe and valid change, and I agree that it'd be nice to include this in 3.8.0/3.7.1, so here's an approval if you'd like to merge today.

**Comment by mimaison** [comment]
cc @soarez as if we want this in 3.7.1, we'll need a new RC.

**Comment by jolshan** [comment]
Apologies for the delay.. Looks like I am two minutes late

**Comment by jolshan** [comment]
But lgtm for what its worth.

**Comment by showuon** [comment]
@edoardocomar , this PR causes compilation error in 3.7 CI. Could you help take a look?
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/3.7/177/pipeline

**Comment by jolshan** [comment]
+1 to Luke's message @soarez did you cherrypick without building 😅  Looks like some classes may have moved around

```

[2024-06-13T12:30:22.380Z] [Error] /home/jenkins/jenkins-agent/workspace/Kafka_kafka_3.7/core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala:25:37: object transaction is not a member of package org.apache.kafka.coordinator

[2024-06-13T12:30:22.380Z] [Error] /home/jenkins/jenkins-agent/workspace/Kafka_kafka_3.7/core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala:26:8: object ServerLogConfigs is not a member of package org.apache.kafka.server.config

[2024-06-13T12:30:22.380Z] [Error] /home/jenkins/jenkins-agent/workspace/Kafka_kafka_3.7/core/src/test/scala/integration/kafka/admin/AdminFenceProducersIntegrationTest.scala:60:15: not found: value ServerLogConfigs
```

**Comment by jolshan** [comment]
Fix here: https://github.com/apache/kafka/pull/16342/files

**Comment by soarez** [comment]
@jolshan 🙈 seems so! Thanks for the fix.

**Pull request merged**